# <b>Exploratory Data Analysis</b>
Here we do some analysis on the data to understand how big the data is, what it looks like and perhaps discover where the errors occur.

## Setup: Imports And Loading Data

In [276]:
import pandas as pd
from pathlib import Path

# path to "raw" data, data from company
dataRaw = Path("../data/raw")

# load all data files
apc = pd.read_csv(dataRaw/'apc_report.csv')
matched_stops = pd.read_csv(dataRaw/'matched_stops.csv', encoding_errors='ignore')
matched_tips = pd.read_csv(dataRaw/'matched_trips.csv')
thoreb = pd.read_csv(dataRaw/'thoreb.csv', encoding_errors='ignore')

C:\Users\Abdullah\AppData\Local\Temp\ipykernel_18188\2822643783.py:11: DtypeWarning: Columns (0: block, 1: diff, 2: journey, 3: line, 4: noJourney, 5: pointID, 6: stoptime, 7: vehicle) have mixed types. Specify dtype option on import or set low_memory=False.
  thoreb = pd.read_csv(dataRaw/'thoreb.csv', encoding_errors='ignore')


## Section 1: APC

Basic information about the file

In [277]:
# shape of the data, (rows, columns)
print("Shape:", apc.shape)

# how many rows is there
print("Number of rows:", apc.shape[0])

# how many columns is there
print("Number of columns:", apc.shape[1])

Shape: (296924, 27)
Number of rows: 296924
Number of columns: 27


Column names in the data

In [278]:
# List column names
apc.columns.tolist()

['fileName',
 'stationName',
 'stationCode',
 'vehicleCode',
 'arrival',
 'departure',
 'boardings',
 'alightings',
 'latitude',
 'longitude',
 'gpsDeviation',
 'operatorCode',
 'isGpsMapped',
 'wheelchairRampAccessCount',
 'isBicycleRackAccessed',
 'area',
 'netstop_latitude',
 'netstop_longitude',
 'pismInformation',
 'boardings_adults',
 'alightings_adults',
 'boardings_children',
 'alightings_children',
 'boardings_wheelchair',
 'alightings_wheelchair',
 'boardings_pram',
 'alightings_pram']

An overview of the data, a few lines from the beginning and a few from the end

In [279]:
#apc.head(5)
#apc.tail(5)
apc

,fileName,stationName,stationCode,vehicleCode,arrival,departure,boardings,alightings,latitude,longitude,...,netstop_longitude,pismInformation,boardings_adults,alightings_adults,boardings_children,alightings_children,boardings_wheelchair,alightings_wheelchair,boardings_pram,alightings_pram
0,NOBINA_3000_20241231_174946_vdv457.raw.gz,NaN,NaN,3000,1/1/2025 0:07,1/1/2025 0:07,0,0,57.111225,12.289127,...,NaN,"pointCode:6555,block:5512,trip:668,tripType:PR...",0,0,0,0,0,0,0,0
1,NOBINA_3000_20241231_174946_vdv457.raw.gz,NaN,NaN,3000,1/1/2025 0:08,1/1/2025 0:08,0,0,57.111214,12.295943,...,NaN,"pointCode:6656,block:5512,trip:668,tripType:PR...",0,0,0,0,0,0,0,0
2,NOBINA_3000_20241231_174946_vdv457.raw.gz,NaN,NaN,3000,1/1/2025 0:09,1/1/2025 0:09,0,0,57.111584,12.301238,...,NaN,"pointCode:6655,block:5512,trip:668,tripType:PR...",0,0,0,0,0,0,0,0
3,NOBINA_3000_20241231_174946_vdv457.raw.gz,NaN,NaN,3000,1/1/2025 0:10,1/1/2025 0:10,0,0,57.111076,12.297543,...,NaN,"pointCode:6656,block:5512,trip:669,tripType:PR...",0,0,0,0,0,0,0,0
4,NOBINA_3000_20241231_174946_vdv457.raw.gz,NaN,NaN,3000,1/1/2025 0:10,1/1/2025 0:11,0,0,57.111470,12.292072,...,NaN,"pointCode:6555,block:5512,trip:669,tripType:PR...",0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
296919,Vaestbo_Taxi_9003_20250107_051551_vdv457.raw.gz,NaN,NaN,9003,7/1/2025 19:39,7/1/2025 19:39,0,0,57.007656,13.241302,...,NaN,"pointCode:4511,block:4173,trip:19,tripType:PRO...",0,0,0,0,0,0,0,0
296920,Vaestbo_Taxi_9003_20250107_051551_vdv457.raw.gz,NaN,NaN,9003,7/1/2025 19:40,7/1/2025 19:40,0,0,57.004128,13.240287,...,NaN,"pointCode:4512,block:4173,trip:19,tripType:PRO...",0,0,0,0,0,0,0,0
296921,Vaestbo_Taxi_9003_20250107_051551_vdv457.raw.gz,NaN,NaN,9003,7/1/2025 19:40,7/1/2025 19:41,0,0,57.001217,13.241087,...,NaN,"pointCode:4502,block:4173,trip:19,tripType:PRO...",0,0,0,0,0,0,0,0
296922,Vaestbo_Taxi_9003_20250107_051551_vdv457.raw.gz,NaN,NaN,9003,7/1/2025 19:41,7/1/2025 19:41,0,0,56.999330,13.238167,...,NaN,"pointCode:4500,block:4173,trip:19,tripType:PRO...",0,0,0,0,0,0,0,0


Information about the data, data types & missing values

In [280]:
apc.info()

<class 'pandas.DataFrame'>
RangeIndex: 296924 entries, 0 to 296923
Data columns (total 27 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   fileName                   296924 non-null  str    
 1   stationName                0 non-null       float64
 2   stationCode                0 non-null       float64
 3   vehicleCode                296924 non-null  int64  
 4   arrival                    296924 non-null  str    
 5   departure                  296924 non-null  str    
 6   boardings                  296924 non-null  int64  
 7   alightings                 296924 non-null  int64  
 8   latitude                   296924 non-null  float64
 9   longitude                  296924 non-null  float64
 10  gpsDeviation               296924 non-null  str    
 11  operatorCode               296924 non-null  str    
 12  isGpsMapped                296924 non-null  bool   
 13  wheelchairRampAccessCount  296924 non-nu

<h3>Missing values:</h1>

Number of nulls per column

In [281]:
apc.isnull().sum()

fileName                          0
stationName                  296924
stationCode                  296924
vehicleCode                       0
arrival                           0
departure                         0
boardings                         0
alightings                        0
latitude                          0
longitude                         0
gpsDeviation                      0
operatorCode                      0
isGpsMapped                       0
wheelchairRampAccessCount         0
isBicycleRackAccessed             0
area                         296924
netstop_latitude             296924
netstop_longitude            296924
pismInformation                 856
boardings_adults                  0
alightings_adults                 0
boardings_children                0
alightings_children               0
boardings_wheelchair              0
alightings_wheelchair             0
boardings_pram                    0
alightings_pram                   0
dtype: int64

Description of the data

In [282]:
apc.describe()

,stationName,stationCode,vehicleCode,boardings,alightings,latitude,longitude,wheelchairRampAccessCount,area,netstop_latitude,netstop_longitude,boardings_adults,alightings_adults,boardings_children,alightings_children,boardings_wheelchair,alightings_wheelchair,boardings_pram,alightings_pram
count,0.0,0.0,296924.000000,296924.000000,296924.000000,296924.000000,296924.000000,296924.0,0.0,0.0,0.0,296924.000000,296924.000000,296924.000000,296924.000000,296924.000000,296924.000000,296924.000000,296924.000000
mean,NaN,NaN,4309.222949,0.349531,0.342401,56.260770,12.572344,0.0,NaN,NaN,NaN,0.342815,0.333308,0.006638,0.009006,0.000010,0.000017,0.000081,0.000084
std,NaN,NaN,1022.862423,1.203051,1.134900,5.632180,1.288397,0.0,NaN,NaN,NaN,1.173637,1.101985,0.126639,0.125759,0.003179,0.004104,0.008990,0.009176
min,NaN,NaN,1.000000,0.000000,0.000000,0.000000,0.000000,0.0,NaN,NaN,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,NaN,NaN,4211.000000,0.000000,0.000000,56.668266,12.454455,0.0,NaN,NaN,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,NaN,NaN,4245.000000,0.000000,0.000000,56.696968,12.808614,0.0,NaN,NaN,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,NaN,NaN,4283.000000,0.000000,0.000000,57.028976,12.890022,0.0,NaN,NaN,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
max,NaN,NaN,9004.000000,64.000000,61.000000,57.364060,16.145123,0.0,NaN,NaN,NaN,64.000000,61.000000,20.000000,10.000000,1.000000,1.000000,1.000000,1.000000


Checking how many busses we have in the apc data, and also the earliest and latest trip.

In [283]:
print("Unique busses:", apc['vehicleCode'].nunique())
print("Earliest arrival:", apc['arrival'].min())
print("Latest arrival:", apc['arrival'].max())

Unique busses: 174
Earliest arrival: 1/1/2025 0:00
Latest arrival: 7/1/2025 9:59


Here we verify that the passenger counts are reliable by checking for negative values and whether the subcategory columns sum up to the total.

In [284]:
# Check for negative values in boarding and alighting columns
passenger_cols = [
    'boardings', 'alightings',
    'boardings_adults', 'alightings_adults',
    'boardings_children', 'alightings_children',
    'boardings_wheelchair', 'alightings_wheelchair',
    'boardings_pram', 'alightings_pram'
]

negative_counts = (apc[passenger_cols] < 0).sum()
print("Negative values per column:")
print(negative_counts)

Negative values per column:
boardings                0
alightings               0
boardings_adults         0
alightings_adults        0
boardings_children       0
alightings_children      0
boardings_wheelchair     0
alightings_wheelchair    0
boardings_pram           0
alightings_pram          0
dtype: int64


In [285]:
# Check if subcategories sum up to the total boardings and alightings
boardings_sum = (
    apc['boardings_adults'] +
    apc['boardings_children'] +
    apc['boardings_wheelchair'] +
    apc['boardings_pram']
)

alightings_sum = (
    apc['alightings_adults'] +
    apc['alightings_children'] +
    apc['alightings_wheelchair'] +
    apc['alightings_pram']
)

boardings_mismatch = (boardings_sum != apc['boardings']).sum()
alightings_mismatch = (alightings_sum != apc['alightings']).sum()

print("Rows where boardings subcategories do NOT sum to total:", boardings_mismatch)
print("Rows where alightings subcategories do NOT sum to total:", alightings_mismatch)

Rows where boardings subcategories do NOT sum to total: 4
Rows where alightings subcategories do NOT sum to total: 3


In [286]:
boardings_mismatch_rows = apc[boardings_sum != apc['boardings']]
alightings_mismatch_rows = apc[alightings_sum != apc['alightings']]

print("Boardings mismatches:")
print(boardings_mismatch_rows[['boardings', 'boardings_adults', 'boardings_children', 'boardings_wheelchair', 'boardings_pram']])

print("\nAlightings mismatches:")
print(alightings_mismatch_rows[['alightings', 'alightings_adults', 'alightings_children', 'alightings_wheelchair', 'alightings_pram']])

Boardings mismatches:
        boardings  boardings_adults  boardings_children  boardings_wheelchair  \
28491           0                 1                   0                     0   
70512           0                 1                   0                     0   
72603           0                 1                   0                     0   
120719          0                 1                   0                     0   

        boardings_pram  
28491                0  
70512                0  
72603                0  
120719               0  

Alightings mismatches:
        alightings  alightings_adults  alightings_children  \
70512            1                  2                    0   
72603            0                  0                    2   
238185           0                  1                    0   

        alightings_wheelchair  alightings_pram  
70512                       0                0  
72603                       0                0  
238185                     

In [287]:
print("isGpsMapped unique values:", apc['isGpsMapped'].unique())
print("gpsDeviation unique values:", apc['gpsDeviation'].unique())

isGpsMapped unique values: [False]
gpsDeviation unique values: <StringArray>
['0 m']
Length: 1, dtype: str


<h3 style="color: red;">Observations:</h3>

1. This columns `(stationName, stationCode, area, netstop_latitude, netstop_longitude)` are empty

2. `pismInformation` contains this info `pointCode,block,trip,tripType,line` that may help us, but there is 856 empty rows in it.

3. apc_report.csv contains data about 174 busses, the data is from 1 to 7 januari 2025.

4. There is no negative values in `boarding or alighting columns`, but there is mismatches in 7 rows.

5. Now we have a good overview of the data and know roughly how we are going to process it.

6. All rows have `isGpsMapped = False` and `gpsDeviation = '0 m'`, meaning no GPS positions were successfully mapped to a known stop. Can coordinates in the APC file be trusted? Or should we use `matched_stops.csv`? 

## Section 2: Matched Stops

Now we analyze matched_stops.csv

Here is basic info about the file

In [288]:
print("Shape:", matched_stops.shape)

print("Number of rows:", matched_stops.shape[0])

print("Number of columns:", matched_stops.shape[1])

Shape: (19154, 25)
Number of rows: 19154
Number of columns: 25


Column names in the data

In [289]:
matched_stops.columns.tolist()

['arrivalTime',
 'compensated.in',
 'compensated.load',
 'compensated.out',
 'compensated.passengerMeters',
 'doors',
 'departureTime',
 'doors.doorNumber',
 'doors.in',
 'doors.out',
 'latitude',
 'longitude',
 'operationDay',
 'optimized.load',
 'optimized.out',
 'optimized.passengerMeters',
 'passengerExchangeSeconds',
 'pointLabel',
 'pointShortLabel',
 'sequence',
 'stopLabel',
 'stopShortLabel',
 'tripNumber',
 'vehicleNumber',
 'optimized.in']

An overview of the data, a few lines from the beginning and a few from the end

In [290]:
matched_stops

,arrivalTime,compensated.in,compensated.load,compensated.out,compensated.passengerMeters,doors,departureTime,doors.doorNumber,doors.in,doors.out,...,optimized.passengerMeters,passengerExchangeSeconds,pointLabel,pointShortLabel,sequence,stopLabel,stopShortLabel,tripNumber,vehicleNumber,optimized.in
0,1/1/2025 6:08:51 AM,0.0,0.0,0.0,0.0,NaN,1/1/2025 6:08:51 AM,NaN,NaN,NaN,...,0.0,0.0,2500-C,2500-C,1.0,Laholm bussterminal,Laholm bussterm,8.196029e+06,138.0,0.0
1,1/1/2025 6:10:13 AM,0.0,0.0,0.0,0.0,NaN,1/1/2025 6:10:38 AM,36.0,0.0,0.0,...,0.0,450.0,2503-B,2503-B,36.0,Laholm Repslagargatan,Laholm Repslagar,1.475285e+08,2484.0,0.0
2,1/1/2025 6:11:11 AM,0.0,0.0,0.0,0.0,NaN,1/1/2025 6:11:54 AM,36.0,0.0,0.0,...,0.0,774.0,2518-B,2518-B,54.0,Laholm Kristianstadsvge,NaN,1.475285e+08,2484.0,0.0
3,1/1/2025 6:12:23 AM,18.0,18.0,0.0,0.0,NaN,1/1/2025 6:13:17 AM,36.0,1.0,0.0,...,7704.0,972.0,2505-B,2505-B,72.0,Laholm Glnninge vrdcen,glnnv,1.475285e+08,2484.0,18.0
4,1/1/2025 6:13:47 AM,18.0,36.0,0.0,0.0,NaN,1/1/2025 6:14:39 AM,36.0,1.0,0.0,...,18468.0,936.0,2507-B,2507-B,90.0,Laholm Altona,Laholm Altona,1.475285e+08,2484.0,18.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19149,1/31/2025 11:27:31 PM,0.0,0.0,0.0,0.0,NaN,1/31/2025 11:27:31 PM,NaN,NaN,NaN,...,1824.0,0.0,2530-B,2530-B,11.0,Laholm Industrigatan,Laholm Industrig,8.607997e+06,204.0,0.0
19150,1/31/2025 11:27:56 PM,0.0,0.0,0.0,0.0,NaN,1/31/2025 11:28:24 PM,36.0,0.0,0.0,...,0.0,504.0,2601-C,2601-C,216.0,Laholm Station,Laholm Station,1.549439e+08,3672.0,0.0
19151,Total,141240.0,487842.0,144396.0,0.0,NaN,NaN,579741.0,10303.0,5056.0,...,300322375.0,43396725.0,NaN,NaN,1891408.0,NaN,NaN,2.524212e+12,59193942.0,148224.0
19152,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Info about the data, data types & missing values

In [291]:
matched_stops.info()

<class 'pandas.DataFrame'>
RangeIndex: 19154 entries, 0 to 19153
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   arrivalTime                  19153 non-null  str    
 1   compensated.in               19152 non-null  float64
 2   compensated.load             19152 non-null  float64
 3   compensated.out              19152 non-null  float64
 4   compensated.passengerMeters  19152 non-null  float64
 5   doors                        0 non-null      float64
 6   departureTime                19151 non-null  str    
 7   doors.doorNumber             16476 non-null  float64
 8   doors.in                     16476 non-null  float64
 9   doors.out                    16476 non-null  float64
 10  latitude                     19152 non-null  float64
 11  longitude                    19152 non-null  float64
 12  operationDay                 19151 non-null  str    
 13  optimized.load             

<h3>Missing values:</h1>

Number of nulls per column

In [292]:
matched_stops.isnull().sum()

arrivalTime                        1
compensated.in                     2
compensated.load                   2
compensated.out                    2
compensated.passengerMeters        2
doors                          19154
departureTime                      3
doors.doorNumber                2678
doors.in                        2678
doors.out                       2678
latitude                           2
longitude                          2
operationDay                       3
optimized.load                     2
optimized.out                      2
optimized.passengerMeters          2
passengerExchangeSeconds           2
pointLabel                         3
pointShortLabel                    3
sequence                           2
stopLabel                          3
stopShortLabel                  1597
tripNumber                         2
vehicleNumber                      2
optimized.in                       2
dtype: int64

Description of the data

In [293]:
matched_stops.describe()

,compensated.in,compensated.load,compensated.out,compensated.passengerMeters,doors,doors.doorNumber,doors.in,doors.out,latitude,longitude,optimized.load,optimized.out,optimized.passengerMeters,passengerExchangeSeconds,sequence,tripNumber,vehicleNumber,optimized.in
count,19152.000000,19152.000000,19152.000000,19152.0,0.0,16476.000000,16476.000000,16476.000000,1.915200e+04,1.915200e+04,19152.000000,19152.000000,1.915200e+04,1.915200e+04,1.915200e+04,1.915200e+04,1.915200e+04,19152.000000
mean,14.749373,50.944236,15.078947,0.0,NaN,70.373999,1.250668,0.613741,1.744813e+03,4.020411e+02,63.069653,8.991228,3.136199e+04,4.531822e+03,1.975155e+02,2.635977e+08,6.181489e+03,15.478697
std,1020.966505,3525.410512,1043.846073,0.0,NaN,4516.291702,80.307029,39.416306,1.207271e+05,2.781804e+04,4364.280813,622.496785,2.170360e+06,3.136041e+05,1.366662e+04,1.823886e+10,4.277109e+05,1071.340849
min,0.000000,0.000000,0.000000,0.0,NaN,18.000000,0.000000,0.000000,0.000000e+00,0.000000e+00,-144.000000,0.000000,0.000000e+00,0.000000e+00,1.000000e+00,8.188850e+06,9.000000e+01,0.000000
25%,0.000000,0.000000,0.000000,0.0,NaN,36.000000,0.000000,0.000000,1.017022e+03,2.340482e+02,0.000000,0.000000,0.000000e+00,4.680000e+02,3.600000e+01,1.474901e+08,2.484000e+03,0.000000
50%,0.000000,0.000000,0.000000,0.0,NaN,36.000000,0.000000,0.000000,1.017076e+03,2.345575e+02,3.000000,0.000000,2.016000e+03,5.760000e+02,9.000000e+01,1.548490e+08,2.484000e+03,0.000000
75%,0.000000,18.000000,0.000000,0.0,NaN,36.000000,0.000000,0.000000,1.017134e+03,2.349356e+02,36.000000,0.000000,1.846800e+04,9.540000e+02,1.440000e+02,1.549074e+08,4.320000e+03,0.000000
max,141240.000000,487842.000000,144396.000000,0.0,NaN,579741.000000,10303.000000,5056.000000,1.670832e+07,3.849945e+06,603955.000000,86100.000000,3.003224e+08,4.339672e+07,1.891408e+06,2.524212e+12,5.919394e+07,148224.000000


Checking how many busses we have in the `matched_stops.csv` data, and also the earliest and latest trip.

In [294]:
# This data file contains just the first week of January 2025, to be compared with apc_report.csv
stops7 = pd.read_csv(dataRaw/'matched_stops7.csv', encoding_errors='ignore')

print("Unique busses:",  stops7['vehicleNumber'].nunique())

print("Earliest arrival:", stops7['arrivalTime'].min())
print("Latest arrival:", stops7['arrivalTime'].max())

Unique busses: 24
Earliest arrival: 1/1/2025 10:08:07 PM
Latest arrival: 1/7/2025 9:58:32 AM


Checking if `matched_stops` can be linked to `apc` data via vehicle number and time range.

In [295]:
apc_vehicles = set(apc['vehicleCode'].unique())
matched_vehicles = set(stops7['vehicleNumber'].dropna().astype(int).unique())

print("Vehicles in APC:", len(apc_vehicles))
print("Vehicles in matched_stops (same week):", len(matched_vehicles))
print("Vehicles in both:", len(apc_vehicles & matched_vehicles))


print("\nMatching vehicle:", apc_vehicles & matched_vehicles)
print("Sample vehicleCode from APC:", list(apc_vehicles)[:10])
print("Sample vehicleNumber from matched_stops:", list(matched_vehicles)[:10])

Vehicles in APC: 174
Vehicles in matched_stops (same week): 24
Vehicles in both: 1

Matching vehicle: {np.int64(5004)}
Sample vehicleCode from APC: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(6), np.int64(4120), np.int64(4121), np.int64(4122), np.int64(3630), np.int64(3631)]
Sample vehicleNumber from matched_stops: [np.int64(262), np.int64(138), np.int64(5004), np.int64(6286), np.int64(278), np.int64(2988), np.int64(5292), np.int64(2484), np.int64(308), np.int64(1080)]


<h3 style="color: red;">Observations:</h3>

1. The `doors` column is completely empty and can be dropped during processing. `doors.doorNumber`, `doors.in` and `doors.out` have 2678 missing values. `stopShortLabel` has 1597 missing values.

2. The file contains stop names (`stopLabel`, `pointLabel`, `stopShortLabel`) and coordinates (`latitude`, `longitude`) which may provide more reliable stop information than the APC coordinates.

3. Only 1 out of 174 vehicles matched. The files cannot be linked via vehicle number.

4. The link between APC and matched_stops needs further investigation, likely via `tripNumber` in matched_stops and the trip information inside `pismInformation` in APC.

## Section 3: Matched Trips

Analysis of matched_trips.csv

Basic info about the file

In [296]:
print("Shape:", matched_tips.shape)
print("Number of rows:", matched_tips.shape[0])
print("Number of columns:", matched_tips.shape[1])

Shape: (1555, 17)
Number of rows: 1555
Number of columns: 17


Column names in the data

In [297]:
matched_tips.columns.tolist()

['data.blockNumber',
 'data.departureTimeSeconds',
 'data.lineShortLabel',
 'data.patternShortLabel',
 'data.tripDeparture',
 'data.tripLabel',
 'data.tripNumber',
 'data.tripShortLabel',
 'data.tripTypeLabel',
 'data.vehicles.lastModified',
 'data.vehicles.locked',
 'data.vehicles.qualityLevel',
 'data.vehicles.stops',
 'data.vehicles.vehicleNumber',
 'links.next',
 'operationDay',
 'links.prev']

Overview of the data, a few lines from the beginning and a few from the end

In [298]:
matched_tips

,data.blockNumber,data.departureTimeSeconds,data.lineShortLabel,data.patternShortLabel,data.tripDeparture,data.tripLabel,data.tripNumber,data.tripShortLabel,data.tripTypeLabel,data.vehicles.lastModified,data.vehicles.locked,data.vehicles.qualityLevel,data.vehicles.stops,data.vehicles.vehicleNumber,links.next,operationDay,links.prev
0,2302-1,22020.0,1201.0,1201-1,1/1/2025 6:07:00 AM,1201-1 1 06:07,8196029.0,1201-1 1 06:07,passenger trip,1/6/2025 9:09:07 AM,True,2.0,/backend-api/report/v1/matched/2025-01-01/trip...,138.0,/backend-api/report/v1/matched/2025-01-01/trip...,1/1/2025 12:00:00 AM,NaN
1,2303-1,22500.0,1201.0,1201-1,1/1/2025 6:15:00 AM,1201-1 1 06:15,8196066.0,1201-1 1 06:15,passenger trip,1/6/2025 9:22:18 AM,True,0.0,/backend-api/report/v1/matched/2025-01-01/trip...,262.0,/backend-api/report/v1/matched/2025-01-01/trip...,1/1/2025 12:00:00 AM,NaN
2,2303-1,25620.0,1201.0,1201-1,1/1/2025 7:07:00 AM,1201-1 1 07:07,8196103.0,1201-1 1 07:07,passenger trip,1/6/2025 9:22:19 AM,True,2.0,/backend-api/report/v1/matched/2025-01-01/trip...,262.0,/backend-api/report/v1/matched/2025-01-01/trip...,1/1/2025 12:00:00 AM,NaN
3,2302-1,26100.0,1201.0,1201-1,1/1/2025 7:15:00 AM,1201-1 1 07:15,8196259.0,1201-1 1 07:15,passenger trip,1/6/2025 9:09:07 AM,True,0.0,/backend-api/report/v1/matched/2025-01-01/trip...,138.0,/backend-api/report/v1/matched/2025-01-01/trip...,1/1/2025 12:00:00 AM,NaN
4,2302-1,29220.0,1201.0,1201-1,1/1/2025 8:07:00 AM,1201-1 1 08:07,8194488.0,1201-1 1 08:07,passenger trip,1/6/2025 9:09:07 AM,True,0.0,/backend-api/report/v1/matched/2025-01-01/trip...,138.0,/backend-api/report/v1/matched/2025-01-01/trip...,1/1/2025 12:00:00 AM,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1550,2306-1,80100.0,1201.0,1201-1,1/31/2025 10:15:00 PM,1201-1 1 22:15,8607561.0,1201-1 1 22:15,passenger trip,2/3/2025 8:17:06 AM,True,0.0,/backend-api/report/v1/matched/2025-01-31/trip...,133.0,/backend-api/report/v1/matched/2025-01-31/trip...,1/31/2025 12:00:00 AM,NaN
1551,2306-1,83220.0,1201.0,1201-1,1/31/2025 11:07:00 PM,1201-1 1 23:07,8607779.0,1201-1 1 23:07,passenger trip,2/3/2025 8:17:06 AM,True,0.0,/backend-api/report/v1/matched/2025-01-31/trip...,133.0,/backend-api/report/v1/matched/2025-01-31/trip...,1/31/2025 12:00:00 AM,NaN
1552,2307-1,83700.0,1201.0,1201-1,1/31/2025 11:15:00 PM,1201-1 1 23:15,8607997.0,1201-1 1 23:15,passenger trip,2/3/2025 8:19:59 AM,True,16.0,/backend-api/report/v1/matched/2025-01-31/trip...,204.0,/backend-api/report/v1/matched/2025-01-31/trip...,1/31/2025 12:00:00 AM,NaN
1553,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Info about the data, data types & missing values

In [299]:
matched_tips.info()

<class 'pandas.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   data.blockNumber             1554 non-null   str    
 1   data.departureTimeSeconds    1553 non-null   float64
 2   data.lineShortLabel          1553 non-null   float64
 3   data.patternShortLabel       1553 non-null   str    
 4   data.tripDeparture           1553 non-null   str    
 5   data.tripLabel               1553 non-null   str    
 6   data.tripNumber              1553 non-null   float64
 7   data.tripShortLabel          1553 non-null   str    
 8   data.tripTypeLabel           1553 non-null   str    
 9   data.vehicles.lastModified   1553 non-null   str    
 10  data.vehicles.locked         1553 non-null   object 
 11  data.vehicles.qualityLevel   1553 non-null   float64
 12  data.vehicles.stops          1553 non-null   str    
 13  data.vehicles.vehicleNumber  

<h3>Missing values:</h1>

Number of nulls per column

In [300]:
matched_tips.isnull().sum()

data.blockNumber                  1
data.departureTimeSeconds         2
data.lineShortLabel               2
data.patternShortLabel            2
data.tripDeparture                2
data.tripLabel                    2
data.tripNumber                   2
data.tripShortLabel               2
data.tripTypeLabel                2
data.vehicles.lastModified        2
data.vehicles.locked              2
data.vehicles.qualityLevel        2
data.vehicles.stops               2
data.vehicles.vehicleNumber       2
links.next                        2
operationDay                      2
links.prev                     1555
dtype: int64

Checking how many busses we have in the matched_tips7.csv data, and also the earliest and latest trip.

In [309]:
trips7 = pd.read_csv(dataRaw/'matched_trips7.csv', encoding_errors='ignore')

print("Unique busses:", trips7['data.vehicles.vehicleNumber'].nunique())
print("Earliest trip:", trips7['data.tripDeparture'].min())
print("latest trip:", trips7['data.tripDeparture'].max())

Unique busses: 12
Earliest trip: 1/1/2025 10:07:00 AM
latest trip: 1/7/2025 9:51:00 AM


Checking if `matched_trips` can be linked to APC data via vehicle number.

In [310]:
apc_vehicles = set(apc['vehicleCode'].unique())
trips_vehicles = set(trips7['data.vehicles.vehicleNumber'].dropna().astype(int).unique())

print("Vehicles in APC:", len(apc_vehicles))
print("Vehicles in matched_trips (same week):", len(trips_vehicles))
print("Vehicles in both:", len(apc_vehicles & trips_vehicles))

Vehicles in APC: 174
Vehicles in matched_trips (same week): 12
Vehicles in both: 0


<h3 style="color: red;">Observations:</h3>

1. `links.prev` is completely empty and can be dropped during processing.

2. `matched_trips.csv` covers a full month. `trips7.csv` was created to match the same period as APC (first week of January 2025), containing 12 unique buses.

3. No vehicles matched between APC and matched_trips, cannot be linked via vehicle number.

4. The link between APC and matched_trips likely needs to go through `data.tripNumber` and the trip information inside `pismInformation` in APC.

## Section 4: Thoreb

thoreb.csv analysis

Basic info about the file

In [315]:
print("Shape:", thoreb.shape)
print("Number of rows:", thoreb.shape[0])
print("Number of columns:", thoreb.shape[1])

Shape: (150003, 13)
Number of rows: 150003
Number of columns: 13


Column names in the data

In [318]:
thoreb.columns.tolist()

['block',
 'day',
 'diff',
 'journey',
 'line',
 'location',
 'noJourney',
 'planned',
 'actual',
 'pointID',
 'stopName',
 'stoptime',
 'vehicle']

Overview of the data, a few lines from the beginning and a few from the end

In [319]:
thoreb

,block,day,diff,journey,line,location,noJourney,planned,actual,pointID,stopName,stoptime,vehicle
0,3515,1/1/2025 12:00:00 AM,-107,868,1306,B,1,12/3/2025 12:01:47 AM,12/3/2025 12:00:00 AM,1,Frennarpsvgen,NaN,4228
1,3518,1/5/2025 12:00:00 AM,31,873,1302,A,1,12/3/2025 11:59:29 PM,12/3/2025 12:00:00 AM,1,Fyllingeskolan,24,4226
2,3522,1/5/2025 12:00:00 AM,-82,716,1312,A,1,12/3/2025 12:01:22 AM,12/3/2025 12:00:00 AM,1,Stora torg,NaN,4206
3,3549,1/7/2025 12:00:00 AM,48,70,1306,B,1,12/3/2025 11:59:12 PM,12/3/2025 12:00:00 AM,1,Folkparken,NaN,4238
4,3555,1/15/2025 12:00:00 AM,123,102,1304,A,1,12/3/2025 11:57:57 PM,12/3/2025 12:00:00 AM,1,Arkitektgatan,NaN,4217
...,...,...,...,...,...,...,...,...,...,...,...,...,...
149998,132,1/22/2025 12:00:00 AM,127.0,2.0,661.0,A,1.0,12/3/2025 6:15:00 AM,12/3/2025 6:17:07 AM,1.0,Sjukhuset,18.0,2003.0
149999,Total,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1854566.0,NaN,NaN,NaN
150000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
150001,Applied filters:\narrivalTime is before 2/1/20...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Info about the data, data types & missing values

In [320]:
thoreb.info()

<class 'pandas.DataFrame'>
RangeIndex: 150003 entries, 0 to 150002
Data columns (total 13 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   block      150002 non-null  object
 1   day        149999 non-null  str   
 2   diff       149999 non-null  object
 3   journey    149999 non-null  object
 4   line       149999 non-null  object
 5   location   149999 non-null  str   
 6   noJourney  149999 non-null  object
 7   planned    149999 non-null  str   
 8   actual     149999 non-null  str   
 9   pointID    150000 non-null  object
 10  stopName   149999 non-null  str   
 11  stoptime   24584 non-null   object
 12  vehicle    149971 non-null  object
dtypes: object(8), str(5)
memory usage: 14.9+ MB


<h3>Missing values:</h1>

Number of nulls per column

In [324]:
thoreb.isnull().sum()

block             1
day               4
diff              4
journey           4
line              4
location          4
noJourney         4
planned           4
actual            4
pointID           3
stopName          4
stoptime     125419
vehicle          32
dtype: int64

Checking how many busses we have in the thoreb7.csv data, and also the earliest and latest trip.

In [334]:
thoreb7 = pd.read_csv(dataRaw/'thoreb7.csv', encoding_errors='ignore')

print("Unique busses:", thoreb7['vehicle'].nunique())

# ??????????????? Not sure ???????????????
print("Earliest trip:", thoreb7['day'].min())
print("Latest trip:", thoreb7['day'].max())

Unique busses: 156
Earliest trip: 1/1/2025 12:00:00 AM
Latest trip: 1/7/2025 12:00:00 AM


Checking if thoreb7 can be linked to APC data via vehicle number.

In [337]:
apc_vehicles = set(apc['vehicleCode'].unique())
thoreb_vehicles = set(thoreb7['vehicle'].dropna().astype(int).unique())

print("Vehicles in APC:", len(apc_vehicles))
print("Vehicles in thoreb7 (same week):", len(thoreb_vehicles))
print("Vehicles in both:", len(apc_vehicles & thoreb_vehicles))

Vehicles in APC: 174
Vehicles in thoreb7 (same week): 156
Vehicles in both: 153


<h3 style="color: red;">Observations:</h3>

1. `stoptime` has a large number of missing values — 125,419 out of 150,003 rows.

2. `thoreb.csv` covers a full month. `thoreb7.csv` was created to match the same period as APC (first week of January 2025), containing 156 unique buses.

3. 153 out of 174 APC vehicles matched with thoreb. Unlike matched_stops and matched_trips.